### PR TITLE
mask_heu_agg

### PR DESCRIPTION
**Pull request by liuzhihaozhengshuai**
# Preamble

For Pull Request initiator, please read the following steps carefully. 请合并请求的发起人仔细阅读下列步骤。
#895 
---

If you are creating a `Pull Request` about listed topic below, please go the the `Preview` tab and click the relative link:
如果您要创建一个有关下列内容的的合并请求，请切换到 `Preview` 面板并点击下方相关的链接:

* Select [Awesome-PETs](?expand=1&template=pull_request_template_pets.md) for files under docs/awesome-pets

---

Select a proper `Pull Request` template can accelarate the review process, thanks for your cooperation.
选择一个正确的`合并请求`模版可以加速审查流程，谢谢您的配合

If your `Pull Request` topic is not amoung the list above, please just delete all prepopulated template content.
如果您的`合并请求`内容不在上述的列表里，请删除提前填入的模版内容。

---

Thanks for your contribution to SecretFlow project
感谢您对于 SecretFlow 项目的贡献

### CODE REVIEW

**Comment by liuzhihaozhengshuai** [comment]
I have read the CLA Document and I hereby sign the CLA

File: secretflow/security/aggregation/experiment/mask_heu_aggregation.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    def add_noisy(self, data, weight=None, mask_location=-2):
+        # 生成噪声self.mask，以及带有噪声的梯度data，他们都乘以了weight
+        # data是梯度向量，是一个list[numpyarray],mask_location表示噪声添加的位置
+        islist = isinstance(data, list)
+        if not islist:
+            data = [data]
+        if len(data) < 2:
+            mask_location = 0
+        data1 = data[mask_location]  # 获取需要保护的位置
```
Comment (zhangxingmeng) [inline secretflow/security/aggregation/experiment/mask_heu_aggregation.py:32]: 参数命名给一个有意义的名字吧

File: secretflow/security/aggregation/experiment/mask_heu_aggregation.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            participants
+        ), 'Should not have duplicated devices.'
+        self._device = device
+        self._participants = set(participants)
+        self.sk_keeper = participants[0]
+        self.evaluators = participants[1:]
+        self.evaluators.append(device)
+        self._fxp_bits = fxp_bits
+        self.mask_location = mask_location
+        # mask_location表示添加噪声的位置，梯度是一个由多个nparray构成的列表，每个nparray表示神经网络中的某一层，该算法向最后一个全连接层添加噪声
```
Comment (zhangxingmeng) [inline secretflow/security/aggregation/experiment/mask_heu_aggregation.py:65]: nparray->ndarray

File: secretflow/security/aggregation/experiment/mask_heu_aggregation.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        data[mask_location] = final_data
+        for i in range(len(data)):
+            data[i] = data[i] * weight
+        if islist:
+            return data
+        else:
+            return data[0]
+
+
+class mask_heu_aggregator(Aggregator):
```
Comment (zhangxingmeng) [inline secretflow/security/aggregation/experiment/mask_heu_aggregation.py:51]: 类名改成驼峰命名吧

File: secretflow/security/aggregation/experiment/mask_heu_aggregation.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        for i, datum in enumerate(data):
+            encrypted_mask[i] = (
+                self._masker[datum.device]
+                .get_mask()
+                .to(
+                    self.heu_device,
+                    config=HEUMoveConfig(heu_dest_party=self._device.party),
+                )
+            )
+        # 对同态加密的mask进行聚合
```
Comment (zhangxingmeng) [inline secretflow/security/aggregation/experiment/mask_heu_aggregation.py:193]: 注释方便的话，可以改成英文

File: secretflow/security/aggregation/experiment/mask_heu_agg_document [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Mask_ Heu_ Aggregation.py presents a new federated learning secure aggregation algorithm, in which participants add noise at some positions of the gradient (the fully connected layer closest to the output layer), perform homomorphic encryption on the noise, and then send the noisy gradient and the noise in the ciphertext state to the server. The server aggregates the noisy gradient in plaintext and the noisy ciphertext in homomorphic addition, Then, the aggregated values of the gradient and plaintext are returned to the participants, who decrypt the noisy aggregated values and subtract them from the noisy gradient aggregated values to obtain a noiseless aggregated result. The specific process is as follows.
```
Comment (zhangxingmeng) [inline secretflow/security/aggregation/experiment/mask_heu_agg_document:1]: 这个可以加到mask_heu_aggregation的docstring里面，
如果是给一个单独的文档，就写的再详细一点吧，如果有reference也贴一下

**Review by zhangxingmeng — DISMISSED** [review]
LGTM

**Comment by 6fj** [comment]
recheck

**Review by zhangxingmeng — APPROVED** [review]
LGTM

**Pull request merged**